### ISSUE TITLE
Sparsity patterns for Jacobian and Hessian computations

### ISSUE THREAD

**Issue by a-jp** [vgvassilev/clad#803]
Based on clad discussion #780.  We assume the current computations of a Jacobian and Hessian are returned as a dense matrix.  We wish to augment the API to allow spare representations of the Jacobian and Hessian to be computed.  This places two requirements:

1. The calculation should return to the user two pieces of information a) in an as yet to be specified sparse data structure, the sparsity structure (the indices of the entries that are non-zero) and b) the values of the non-zero entries (see 2) in a compressed format.    
2. The non-zero entries computed and returned should be structurally non-zero as a function of the differentiation, not due to initial values or constants.  Said another way, the zero entries that are excluded must be defined to be *always* zero, and not just zero due to an specific initial values.

As a check on the results outputted from 1 and 2, one should be able to reconstruct a dense matrix from the two pieces of information provided by 1. with additional zeros where needed.  For a sanity check, one would then expect the recombination of the sparsity structure, and the compressed non-zero entries into a dense matrix to be identical (within machine precision) to the currently computed dense matrices computed by calls to the Jacobian and Hessian routines.

It's for library developers to comment on the best mechanics for integration into clad, a small amount of the discussion is contained in the initial discussion #780.  It's possible that something is to be learned from other works such as [CasADi](https://web.casadi.org/docs/#the-sparsity-class) although I neither use or have any affiliation with that codebase.  Likewise, sparse Hessians and Jacobians are well accommodated for, as noted in the original discussion, in [CppAD](https://cppad.readthedocs.io/latest/).

**Comment by bradbell** [comment]
You may want to look at the following:
https://arxiv.org/abs/2111.05207

**Comment by a-jp** [comment]
Hi, I was just checking in to see if you still felt v1.6 was a probable version for this capability to be released?

Many thanks
Andy

**Comment by vgvassilev** [comment]
@vaithak probably can comment on his plans about this.

**Comment by bradbell** [comment]
Perhaps it is possible to use some of the sparstiy code in CppAD to reduce the work ?

**Comment by vaithak** [comment]
> Perhaps it is possible to use some of the sparstiy code in CppAD to reduce the work ?

That would be really helpful, thanks @bradbell. I will try going through the paper along with the implementation in CppAD. What would be the good medium of communication to discuss any doubts I encounter, either in the paper or the implementation?

**Comment by bradbell** [comment]
I thought that perhaps we could work together to define an interface for a sparsity tool that could be used by any AD package. The input would be things like the DAG  for the program. The output would be things like the sparsity pattern and splitting for which forward or backward derivatives can be computed during the same pass.

**Comment by vaithak** [comment]
Seems like a good idea. 
@vgvassilev, what do you think about this?

**Comment by vgvassilev** [comment]
That’d make sense to me.

**Comment by bradbell** [comment]
@a-jp send me an e-mail message

**Comment by a-jp** [comment]
> I thought that perhaps we could work together to define an interface for a sparsity tool that could be used by any AD package. The input would be things like the DAG  for the program. The output would be things like the sparsity pattern and splitting for which forward or backward derivatives can be computed during the same pass.

@vgvassilev and @vaithak just to follow up on my own post above, for which I would welcome your comments. To be clear there are two interfaces. The one quoted here should deal with the implementation. My post above relates to the user experience and interface. One of the aspects of Clad that appeals to me is abstraction. I do not want to know about a DAG as a user. I'd want to call 'hessian' or 'sparse_hessian' and that be the only distinction at the user level. Happy with setup/init calls etc. One aspect of other libraries that while being very powerful is it's almost impossible to simply transition between the two interfaces without extensive understanding of what's going on. As a user this is not a requirement that needs to be enforced. As I say, happy to chat about interfaces from a user point of view and this would allow developers to plan in the background for the transition between the two schemes.

**Comment by bradbell** [comment]
@a-jp I am not thinking of the user using a DAG, rather clad passing some sort of DAG to a low level library that does the sparsity calculations for it. We are discussing this on the clad channel of the discord server. Perhaps @vgvassilev can set you up with an invite to that server.

**Comment by a-jp** [comment]
Hi,

Wanted to understand a little bit more about the plans for this in terms of roadmap/timeframe/blockers. Can you share?  Thanks

**Comment by vgvassilev** [comment]
> Hi,
> 
> Wanted to understand a little bit more about the plans for this in terms of roadmap/timeframe/blockers. Can you share? Thanks

Honestly I am a bit lost what exactly the implementation should do. The way I understand the current feature request is essentially to provide a way to "select" and "deselect" directions from the hessians. That would allow the user to express the computations in a way which would reduce computations.

Assuming my assessment is correct then Clad tries to tackle this by implementing to-be-recorded analysis and activity analyses (still not complete) as described by [Tapenade paper](https://www.researchgate.net/publication/258023750_The_Tapenade_Automatic_Differentiation_Tool_Principles_Model_and_Specification).

Allowing the users to automatically select directions is currently challenging to implement. That is, I do not see who to implement it without a lot of infrastructure. Consider,

```cpp
double add(double a, double* b) {
    return a + b[0];
}
double fn17 (double x, double* y) {
    x = add(x, y);
    x = add(x, &x);
    return x;
}
int main() {
auto fn17_grad_0 = clad::gradient(fn17, "x"); // select only one direction...
}
```

then the generated code looks like:

```cpp
void add_pullback(double a, double *b, double _d_y, double *_d_a);
void add_pullback(double a, double *b, double _d_y, double *_d_a, double *_d_b);
void fn17_grad_0(double x, double *y, double *_d_x) {
    double _t0 = x;
    x = add(x, y);
    double _t1 = x;
    x = add(x, &x);
    *_d_x += 1;
    {
        x = _t1;
        double _r_d1 = *_d_x;
        *_d_x = 0.;
        double _r1 = 0.;
        add_pullback(x, &x, _r_d1, &_r1, &*_d_x);
        *_d_x += _r1;
    }
    {
        x = _t0;
        double _r_d0 = *_d_x;
        *_d_x = 0.;
        double _r0 = 0.;
        add_pullback(x, y, _r_d0, &_r0);
        *_d_x += _r0;
    }
}
void add_pullback(double a, double *b, double _d_y, double *_d_a) {
    *_d_a += _d_y;
}
void add_pullback(double a, double *b, double _d_y, double *_d_a, double *_d_b) {
    {
        *_d_a += _d_y;
        _d_b[0] += _d_y;
    }
}
```

This means that now we have to generate pullback operators with reduced set of parameters and computation at the level of planning the differentiation steps. In cases where `x` is assigned to something else we need to track these things too.

Am I making any sense?

**Comment by bradbell** [comment]
Actually, the user should not have to do anything at all, except say that a sparse instead of dense Jacobian should be computed. In CppAD, all the sparse routines require an automatic sparsity computation (which requires time and memory) before the derivative can be computed.

Note that a sparse Hessians can be computed by first computing the gradient and then computing the sparse Jacobian of the gradient. So for a first cut, I would just consider sparse Jacobians. Now suppose that

f_0 (x) = x_0
f_1(x) = x_1

Then the sparse Jacobian can be computed using the directional derivative of f in one direction; e.g. (1,1),  and using the fact that the partial of f_0 w.r.t x_1  and partial of f_1 w.r.t. x_0 is zero.

**Comment by vaithak** [comment]
What I understand from the discussion above is that there should be a way to compute and return a sparse Jacobian / Hessian matrix. For example: `clad::sparse_jacobian(f)`;

This should ensure that only non-zero entries of the Jacobian matrix should be computed.
This will ensure that the returned result uses less memory - because of the dense vs sparse matrix. Also, the computation will be less, as only a few partial derivatives are required to be computed.

The implementation will require an activity analysis like pass over a possibly vector-valued function to recognize which outputs depend on which subset of inputs. The implementation can be referred from the paper that @bradbell shared.

The user won't have to specify the direction of computation. If clad tells them that the (i,j) element of the Jacobian is non-zero, it means `(df_i/dx_j)` would be non-zero, i.e. partial derivative of ith output of `f` w.r.t jth input of `f`.

The generated function can look something like:
```cpp
void f_sparse_jacobian(double x, double y, SparseMatrix<double> &Jacobian>);
```
For this, Clad will need a custom/eigen-like sparse matrix representation (https://eigen.tuxfamily.org/dox/group__TutorialSparse.html) .

**Comment by a-jp** [comment]
Yep. This is consistent with what I was requesting and said in the accompanying discussion to this issue. 

When I say that I think it's reasonable to only request sparse jacobians and sparse hessians that algebraically have zero entries, does that makes sense? I'm happy to clarify that.

**Comment by vgvassilev** [comment]
> Yep. This is consistent with what I was requesting and said in the accompanying discussion to this issue.
> 
> When I say that I think it's reasonable to only request sparse jacobians and sparse hessians that algebraically have zero entries, does that makes sense? I'm happy to clarify that.

Please clarify that too. That issue would land on a student's plate and the more clear are we about what's required the better.

**Comment by a-jp** [comment]
So I mean that if a derivative resulted in constants only in the term for that entry. Then we'd want to undertake this sparsity calculation, which I assume is happening anyway, without needing to know the values of the constants. It's clear that depending on their values the entry in the Jacobian or hessian may too be then zero, but we live with the fact that we don't want to do this at runtime so constants are unknown. We're only looking for entries that are algebraically zero to be defined as sparse.  Have I confused issues or is that clear, and  that was what was going to happen anyway?

Thanks

**Comment by vaithak** [comment]
Hi @a-jp,
Just to clarify your above point with an example.
Let's say `f(x, y) = [x^2, y^2]`, so 2 inputs, 2 outputs.

This implies the Jacobian will be a 2x2 matrix, and have the following form:
```
[[2x, 0],
[0, 2y]]
```
This means that Clad should infer that only diagonal entries (algebraically non-zero ones) need to be computed.  
Although, for the initial values (x = 0), and/or (y = 0), even these diagonal entries may be 0, but that is only for a subset of initial values, and may only be known at runtime. So, it is ok that Clad still computes them

**Comment by a-jp** [comment]
Yes that's a much cleaner/nicer way of saying the same thing as me 😀

**Comment by bradbell** [comment]
A sparsity pattern should identify the entries that may be non-zero. I call one pattern more efficient than another if it has fewer possibly non-zero entries.

**Comment by vgvassilev** [comment]
@a-jp, iirc, @ovdiiuv will start looking into this issue shortly. Would you be interested to organize a call?

**Comment by a-jp** [comment]
Yes, that would be interesting

**Comment by a-jp** [comment]
No, I am not a user of discord

**Comment by ovdiiuv** [comment]
> No, I am not a user of discord

Let's move this discussion to email. Could you please contact me at kaptain.vrungel[at]gmail.com, so that we could organize a call.
